### PR TITLE
Idr prototype

### PR DESCRIPTION
**Pull request by will-moore**
Update to Gallery to allow searching & filtering of top-level Projects & Screens and a nicer
layout showing 'categories' of interest, selected by querying Map-Annotations.
See https://github.com/openmicroscopy/design/issues/100

Can be tested at http://web-dev-merge.openmicroscopy.org/gallery/

For screenshots etc see the design issue link above.

TODO:
 - [x] Fix Key-Value auto-complete for /cells and /tissues (broken since I started using config of 'categories' to define studies in "tissue").
    - The ```cells``` super_category is defined as "other", meaning it includes everything that is not in ```tissue``` category. Q: Is this strategy the way that we want to define 'cells' vv 'tissue'??
 - [x] Check that all URLs generated in JS are based on 'base_url'
 - [x] Move hard-coded ```FILTER_KEYS``` to config
 - [x] Same for mapr config - list subset of mapr keys to appear in the UI
 - [x] Config for title "Welcome to IDR"
 - [ ] Customise UI: Header Icon, other html? links etc? See https://github.com/ome/omero-gallery/issues/26
 - [ ] Browser support? Not IE 10 unless we transpile JavaScript. Need build step, e.g. webpack See https://github.com/ome/omero-gallery/issues/27
- [ ] Fix listing of multiple Screens/Projects for a single idr00xx Study, and do this in a non IDR-specific way? See https://github.com/ome/omero-gallery/issues/28
 - [x] Don't try ```render_image/ID/``` for Big Images. Maybe load sizeX and sizeY when we get ```study/image/``` JSON?
 - [x] Don't use ```/idr/``` directory for static files & templates
 - [x] Handle duplicate keys e.g. FRAP in https://idr.openmicroscopy.org/webclient/?show=project-753 for auto-complete (and filtering).
 - [x] Show "Most Recent" category. Update category labels as suggested by Frances"
 - [x] Auto-complete show on input focus()
 - [x] Search by idr00xx number
 - [x] Error handling on all ```fetch()``` calls.
 - [x] Allow 'Enter' to goto /search without having to pick auto-complete option
 - [x] Always show links to "Cells" and "Tissue" in the header
 - [x] If filter gives 0 results, say "Try your search on ***-IDR?"
 - [ ] Case sensitive option for IDR?
 - [x] Show auto-complete options for MAPR without user needing to type, e.g. for Cell Lines (as in idr mapr)
 - [x] Front page of /cells = "Welcome to Cell-IDR" and of /tissue = "Welcome to Tissue-IDR".
 - [x] Show spinner when during auto-complete.
 - [x] Study Title config to use Map-Annotation for Study/Publication Title

Need Help:
 - [x] Config web-dev-merge to allow to connect to IDR "public", "public"
 - [x] What "categories" do we want? Does current query support what we need? Needs curation.
 - [ ] Use 'live' mapr config loaded from IDR. First requires https://github.com/ome/omero-mapr/pull/46 to be merged and deployed somewhere. If not IDR then need a temp config to point elsewhere.
 - [ ] Also need to deploy, test and merge https://github.com/ome/omero-mapr/pull/45 so we can link to Study within mapr.
 - [x] Test with various configs. *At least*: current working configs vv None, but more is better.
 - [x] Move configs from default values in ```gallery_settings.py``` to deployment CI (and write docs)

### CODE REVIEW

**Comment by will-moore** [comment]
Summary of Study Types: count (NB: duplicate studies e.g. idr0038 are counted multiple times).
```
3D-tracking of tagged chromatin loci: 1
4D timelapse of in vivo cells: 1
5D light sheet and confocal microscopy imaging of mouse zygotes: 1
Confocal laser scanning microscopy of cytoskeletal systems: 1
Quantitative protein localization using FCS-calibrated 3D time-lapse imaging: 1
high content analysis of cells: 1
high content screen: 37
high resolution electron microscopy: 1
histology: 2
imaging mass cytometry: 1
in situ hybridization: 1
light sheet fluorescence microscopy of TARDIS registered embryos: 1
light sheet fluorescence microscopy of organoids: 4
light sheet microscopy of zebrafish tailbud: 1
microscopy assay: 4
protein localization: 1
protein localization using 3D-SIM: 1
protein localization using dSTORM: 1
timelapse microscopy: 1
```

**Comment by will-moore** [comment]
Some potential categories: IDs are idr numbers.
```
"time-lapse": [41, 27, 26, 40],
"lightsheet": [51, 45,  44, 38],
"protein localization": [43, 41, 23, 21, 19],
"histology": [42, 18, 43],
"yeast": [3, 4, 7, 11, 27, 40, 47],
"human nuclear stain": [16, 9, 6, 17, 30, 36, 10]
```

**Comment by manics** [comment]
If I enter a search term which returns some results, then delete the search term, the results are removed but the count `Showing 37 of 63 studies` remains unchanged:
Screen Shot 2019-05-03 at 16 47 19

**Comment by francesw** [comment]
Page looks good!  

Here are some comments.  

Heading: Welcome to the Image Data Resource (IDR).  
Remove “Cells” and “Tissue” from images and have “Cell-IDR” and “Tissue-IDR” above images?  

Drop-down for search/filter:
•	Remove “Study Attributes” and “Image Attributes”?  Note: Organism appears twice.  
•	Search/filter list (in alphabetical order?) = IDR Accession (default?), Imaging Method, Organism, Cell Line, Gene, siRNA, Antibody, Compound, Phenotype, Publication Author, Publication Title.  
•	For publication authors, include last authors as well (i.e. the PI).  
•	When typing in search box, list of suggestions should reflect what has been typed in the search box (i.e. when “t” is in Imaging Method and “imaging mass cytometry” should not be at the top of the list).  
•	Possibly not have the space/gap between the filter and search box?  

Question:
•	Is it possible to type in a search term without selecting a filter?  

Categories suggestions: Time-lapse imaging, Light sheet fluorescence microscopy, High-content screening, Digital pathology imaging (instead of histology?), Protein localization studies, Human cell line studies, Yeast studies.  

•	Make scroll bar more obvious for categories with more than 4 results?  
•	Category links should be directed to new page containing the thumbnails for that category (i.e. light sheet = 7)?  
•	Only show categories with results for cell and tissue IDR pages.

**Comment by manics** [comment]
When the autocomplete dropdown contains long lines (e.g. study titles) the wrapping means the second line is indented to the left of the first:
Screen Shot 2019-05-03 at 17 00 30

Is it possible to get the second line to be indented to the right instead by several spaces, I think this would be easier to read.

**Comment by manics** [comment]
Suggestion: Add idr accession number as an option, allow entries such as `idr0052` `0052`, `52`

**Comment by will-moore** [comment]
New to test today:
 - Switching the "Search By" select to various "Study Attributes" options auto-populates the auto-complete so you can see example values, instead of having a blank input to start typing in.
 - Auto-complete options show number of matches in brackets for "Study Attributes" and are also sorted to show results that begin with the input-text first.
 - Filtering by MAPR - results now show up to 4 thumbnails per study where the images actually match the query (previously the thumbnail for each study was not related to the MAPR query)
 - Scrollbars on the category sections should now be obvious without having to try scrolling.

**Comment by manics** [comment]
Suggestion: have a link to go back to the default view of everything

Minor UI bug (not a blocker for merging):
`light sheet fluorescence microscopy of TARDIS registered embryos` flips between between split over one line and two depending on whether it's highlighted:

Screen Shot 2019-05-07 at 11 05 54
Screen Shot 2019-05-07 at 11 06 01

The lack of default suggestions for the `Image Attributes` is unexpected after the `Study Attributes`, but if this is going too involve some mapr or caching optimisation this can go into a new issue.

File: omero_gallery/static/gallery/idr/model.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+
+StudiesModel.prototype.loadStudiesMapAnnotations = function loadStudiesMapAnnotations(callback) {
+  let url = this.base_url + "/webclient/api/annotations/?type=map";
+  let data = this.studies
+    .map(study => `${ study['@type'].split('#')[1].toLowerCase() }=${ study['@id'] }`)
+    .join("&");
+  url += '&' + data;
+  // Cache-buster. See https://trello.com/c/GpXEHzjV/519-cors-access-control-allow-origin-cached
+  url += '&_=' + Math.random();
```
Comment (manics) [inline omero_gallery/static/gallery/idr/model.js]: Can be removed when https://github.com/IDR/deployment/pull/176 is deployed to test64

File: omero_gallery/static/gallery/idr/search.js [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+
+// ------------ Handle MAPR searching or filtering --------------------- 
+
+function filterStudiesByMapr(value) {
+  let configId = document.getElementById("maprConfig").value.replace("mapr_", "");
+  let url = `${ BASE_URL }/mapr/api/${ configId }/?value=${ value }`;
+  // Cache-buster. See https://trello.com/c/GpXEHzjV/519-cors-access-control-allow-origin-cached
+  url += '&_=' + Math.random();
```
Comment (manics) [inline omero_gallery/static/gallery/idr/search.js]: Can be removed when https://github.com/IDR/deployment/pull/176 is deployed to test64

File: omero_gallery/static/gallery/idr/search.js [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+        $.ajax({
+            dataType: "json",
+            type : 'GET',
+            url: url,
+            data: {
+                value: case_sensitive ? request.term : request.term.toLowerCase(),
+                query: true,
+                case_sensitive: case_sensitive,
+                "_": Math.random()     // cache-buster
```
Comment (manics) [inline omero_gallery/static/gallery/idr/search.js]: Will this be fixed when https://github.com/IDR/deployment/pull/176 is deployed to test64 or is it a separate browser caching workaround?

File: omero_gallery/static/gallery/idr/index.js [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+        $.ajax({
+            dataType: "json",
+            type : 'GET',
+            url: url,
+            data: {
+                value: case_sensitive ? request.term : request.term.toLowerCase(),
+                query: true,
+                case_sensitive: case_sensitive,
+                '_': Math.random,    // cache-buster
```
Comment (manics) [inline omero_gallery/static/gallery/idr/index.js]: Will this be fixed when https://github.com/IDR/deployment/pull/176 is deployed to test64 or is it a separate browser caching workaround?

**Comment by joshmoore** [comment]
@will-moore : would it be possible to migrate the bigger items from your TODO list to separate issues? For discussions at this point, we'd have to reference the line from your description and then comment. It will make discussions hard to follow. From my side, I'm for getting this PR in as quickly as possible, and then following up with other targeted PRs. If you're worried about having conflicts with yourself, then perhaps we do that via milestones (this being M1?) rather than a PR per TODO. So if you had 5 issues for a given milestone, then in the description you could have:

```
 - [x] fix #A
 - [x] fix #B
```
and merging that new PR would close all the tickets.

**Comment by will-moore** [comment]
Request to filter by License https://lists.openmicroscopy.org.uk/mailman/private/idr-submission/2019-May/002307.html added in e1c3bcd

**Comment by chris-allan** [comment]
Chrome on Windows 10. Just over 1080p browser window size. Default font scaling.

Few pieces of feedback for the landing page:

1. Getting a series of JavaScript errors in the console on initial page load:
```No Dataset in Project! {meta: {…}, data: Array(0)}
model.js:312 Uncaught (in promise) TypeError: Cannot read property 'json' of undefined
    at model.js:312
(anonymous) @ model.js:312
Promise.then (async)
loadImage @ model.js:313
(anonymous) @ categories.js:223
loadStudyThumbnails @ categories.js:217
render @ categories.js:173
(anonymous) @ categories.js:264
(anonymous) @ model.js:191
Promise.then (async)
loadStudiesMapAnnotations @ model.js:162
(anonymous) @ model.js:144
Promise.then (async)
loadStudies @ model.js:124
(anonymous) @ categories.js:259
model.js:305 No Dataset in Project! {meta: {…}, data: Array(0)}
model.js:312 Uncaught (in promise) TypeError: Cannot read property 'json' of undefined
    at model.js:312
```
2. I've got a pretty fast connection, am ~20ms from idr.openmicroscopy.org and the page takes over 10 seconds to fully load. Might be an idea to batch thumbnail calls via `get_thumbnails`. The page makes 50+ individual requests for thumbnails at the moment and this will only get worse over time.
3. Not confining the author lists and not having some line breaks in the descriptions creates some pretty wacky visualization conditions:
image
image
4. The scaling of the thumbnail strips is creating some weird excessive whitespace in the horizontal direction. You can scroll well beyond any thumbnails:
image
5. Yeast Studies scrollbar/header overlap:
image

And for search:

1. If idr0018 is matched by a search, such as "microscopy", it is missing some of its metadata:
image
2. Feels kind of weird to have a search results grid that is only two studies wide
image

**Comment by will-moore** [comment]
@chris-allan Thanks for the feedback. I've fixed the console errors and search results is now 3 columns wide. The other css fixes I have put on hold since the whole page is likely to get a facelift and I was already spending too much time trying to design and fix those.

**Comment by chris-allan** [comment]
:+1:

Already looking pretty solid and adept at what it's trying to do. Is your github.io up to date with the latest code? Just noticed there are still console errors there.

**Comment by will-moore** [comment]
@sbesson @francesw As noted by @chris-allan, ```idr0018``` is missing ```Publication Title``` both in the Project Description and in the linked Map Annotation. I'm getting this from the Description to show in the UI like this:
```
  let titleRe = /Publication Title\n(.+)[\n]?/
  let match = titleRe.exec(desc);
  let title = match ? match[1] : '';
```
Do we want to add some title to this study to "fix" the UI?

**Comment by will-moore** [comment]
@chris-allan I hadn't pushed the console error fixes to github.io. But did it just now.

**Comment by chris-allan** [comment]
Great. Definitely no error messages now.

**Comment by will-moore** [comment]
To test last commit (batch loading of thumbnails), first login to "idr" as "public/pubilc" at http://web-dev-merge.openmicroscopy.org/webclient/login/ then go to:
http://web-dev-merge.openmicroscopy.org/gallery/
You should see thumbnails loading much faster than before. Instead of making 2 JSON calls to get the image ID then the loading of the thumb itself (3 calls for each image) we now load the Image IDs and thumbnails for 10 studies/thumbnails at once. (30 x fewer calls). 
URLs like http://web-dev-merge.openmicroscopy.org/gallery/api/thumbnails/?project=51&project=101&project=402&screen=1603&screen=1602&screen=1601&screen=1551&screen=1501&screen=201&screen=202 are used to get thumbnails AND image IDs (needed for the viewer links) in one shot.

**Comment by chris-allan** [comment]
Looking now at the aggregate changes in `urls.py` and `views.py` I can see us slowly going down the road of adding some pretty fundamental API to an OMERO.web plugin. Representative thumbnails for containers is something that has been in the background for a long time. Glencoe did something similar for the JCB DataViewer in what we called the “featured” image via a specifically namespaced BooleanAnnotation and a fallback to the first image if missing.

I’d like to propose that we iterate on this API using https://github.com/glencoesoftware/omero-ms-thumbnail. Doing so is very much in the spirit of trying to avoid the deployment idiosyncrasies mentioned above, reinforces a separation of concerns, and also helps avoid the need for a full OMERO.web deployment being required to contribute to this plugin.

/cc @kkoz, @emilroz

**Comment by sbesson** [comment]
Re-tested the latest version of this code on the deployed CI this morning. Most of the bug fixes discussed above have been addressed. I will capture some errors as issues.

As noted in https://github.com/ome/omero-gallery/pull/25#issuecomment-491765164, the size of this PR is starting to be big enough that reviewing the code and the comments becomes quickly impractical. In line with various conversations, I think we all agree that we have now reached the end of the first phase which was dedicated to building this UI.

Proposing to tag this initial effort as 3.2.0a1 and start working on the next steps in smaller steps, releasing the app in milestones as needed:

1. make the changes to deploy this application into a testing resource in preparation of production deployment. This might require a couple of additional cleanup PRs to remove hard-coded configuration, enable thumbnail viewing without the need to log in etc.
2. iterate over the UI issues (CSS & other bugs) as well as the metadata that should be added to the studies to make the UI functional
3. once deployed, schedule some performance testing of the new UI
4. investigate the migration of the required thumbnail API calls to a micro-service as suggested in https://github.com/ome/omero-gallery/pull/25#issuecomment-494756166. This might also be highlighted by the load testing.
